### PR TITLE
1882: Keycloak display identity providers linked with the user on the login password page

### PR DESCRIPTION
**Pull request by collinpreston**
# What are the relevant tickets?
https://github.com/mitodl/ol-infrastructure/issues/1882

# Description (What does it do?)
A. Identifies which IDp's a user is linked with on the login-password page when the username of the user is known.  The username of the user can be known when the browser-login flow contains the Username Form before the Open Learning Username Password Form, for example:
Screenshot 2023-11-06 at 11 08 38 AM

B. Adds in the custom Open Learning Keycloak theme that used to exist here (https://github.com/mitodl/ol-keycloak-customization).  The theme is included into this repo so that custom extensions and the theme are compiled into jar files at the same time.  This will hopefully make managing and deploying Keycloak customizations easier and only require a single repository for both theme and extensions.

# Screenshots (if appropriate):
**View for a user linked with a single IDP**
Screenshot 2023-11-06 at 11 04 43 AM

**View for a user not linked with any IDP**
Screenshot 2023-11-06 at 11 07 48 AM

# How can this be tested?

1. Follows the steps outlined in the PR's README in order to integrate the PR changes into a local instance of Keycloak.
2. Create 2 dummy Identity Providers. (details don't matter, just ensure the names are different).
3. Create a user and manually link them to one of the identity providers from step 2 using the Keycloak Admin Console.  You can enter anything into the IDP link form.
4. Set the login theme to use the custom OL theme included in this PR.
5. Create a browser login flow that matches screenshot in the "Description" section of this PR.
6. Logout of Keycloak and attempt to log back in using the user from step 3.
7. Verify that the username-password page displays the IDP which the user is connected with and the IDP the user is not connect with (should look like the "View for a user linked with a single IDP" screenshot).
8. Log back into the Keycloak admin console.
9. Link the user from step 3 to both IDPs from step 2.
10. Perform step 6.
11. Verify that the username-password page displays the IDP which the user is connected with.

# Notes
This PR does not include unit tests.  Keycloak does not provide an easy API to build unit tests on top of.  Because of that, there will need to be considerable work done to build infrastructure into this repository in order to support unit tests.  I will create an issue for this work to be done in the future.

### CODE REVIEW

File: ol-keycloak/ol-spi/pom.xml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.example</groupId>
```
Comment (shaidar) [inline ol-keycloak/ol-spi/pom.xml:7]: Should that be `org.ol` or `org.mit.ol`?

File: ol-keycloak/ol-spi/src/main/java/authentication/authenticators/browser/UsernamePasswordForm.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import java.net.URI;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class UsernamePasswordForm extends org.keycloak.authentication.authenticators.browser.UsernamePasswordForm {
+
+    @Override
+    public void authenticate(AuthenticationFlowContext context) {
+        if (context.getUser() != null) {
+            final RealmModel realm = context.getRealm();
```
Comment (shaidar) [inline ol-keycloak/ol-spi/src/main/java/authentication/authenticators/browser/UsernamePasswordForm.java:20]: What happens if a user does not have an account and types in their username, what Realm will this return?

Comment (collinpreston) [reply ol-keycloak/ol-spi/src/main/java/authentication/authenticators/browser/UsernamePasswordForm.java:20]: The user will not reach this screen with the flow shown in the PR description

**Comment by collinpreston** [comment]
This will require a Concourse update to build and incorporate this repositories theme and extension jar files.  It will also require a Pulumi update to configure the login flow described in the PR description.

File: ol-keycloak/ol-spi/pom.xml (hunk 1 above)
Comment (collinpreston) [reply ol-keycloak/ol-spi/pom.xml]: Updated to `edu.mit`

File: ol-keycloak/ol-spi/pom.xml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>edu.mit</groupId>
```
Comment (blarghmatey) [inline ol-keycloak/ol-spi/pom.xml:7]: Should this be `edu.mit.openlearning`?

**Review by shaidar — APPROVED** [review]

**Pull request merged**